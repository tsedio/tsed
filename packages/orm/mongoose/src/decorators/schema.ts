import {decoratorTypeOf, StoreMerge, useDecorators} from "@tsed/core";
import {registerProvider} from "@tsed/di";
import {Property} from "@tsed/schema";
import {SchemaTypeOptions} from "mongoose";
import {MONGOOSE_SCHEMA} from "../constants";
import {MongooseSchemaOptions} from "../interfaces";
import {getSchema, getSchemaToken} from "../utils/createSchema";

/**
 * Define a class as a Mongoose Schema ready to be used to compose other schemes and models.
 *
 * ### Example
 *
 * ```typescript
 * @MongooseSchema()
 * export class EventSchema {
 *   @Property()
 *   field: string;
 * }
 * ```
 *
 * ### Options
 *
 * - `schemaOptions` (mongoose.SchemaOptions): Option to configure the schema behavior.
 *
 * @param {MongooseSchemaOptions | undefined} options
 * @returns {(target: any) => void}
 * @decorator
 * @mongoose
 * @property
 * @class
 */
export function Schema(options?: MongooseSchemaOptions): (target: any) => void;
export function Schema(definition: SchemaTypeOptions<any>): Function;
export function Schema(options: MongooseSchemaOptions | SchemaTypeOptions<any> = {}) {
  return (...parameters: any[]) => {
    switch (decoratorTypeOf(parameters)) {
      case "property":
        return useDecorators(Property(), StoreMerge(MONGOOSE_SCHEMA, options))(parameters[0], parameters[1], parameters[2]);

      case "class":
        const {token} = getSchemaToken(parameters[0], options);

        registerProvider({
          provide: token,
          deps: [],
          useFactory() {
            return getSchema(parameters[0], options as any);
          }
        });
        break;
    }
  };
}

/**
 * Define a class as a Mongoose Schema ready to be used to compose other schemes and models.
 *
 * ### Example
 *
 * ```typescript
 * @MongooseSchema()
 * export class EventSchema {
 *   @Property()
 *   field: string;
 * }
 * ```
 *
 * ### Options
 *
 * - `schemaOptions` (mongoose.SchemaOptions): Option to configure the schema behavior.
 *
 * @param {MongooseSchemaOptions | undefined} options
 * @returns {(target: any) => void}
 * @decorator
 * @mongoose
 * @property
 * @class
 */
export function MongooseSchema(options?: MongooseSchemaOptions): (target: any) => void;
export function MongooseSchema(definition: SchemaTypeOptions<any>): Function;
export function MongooseSchema(options: MongooseSchemaOptions | SchemaTypeOptions<any> = {}) {
  return Schema(options as any);
}
