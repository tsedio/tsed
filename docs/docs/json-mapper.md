---
meta:
  - name: description
    content: Documentation over JsonMapper and decorators provided by Ts.ED framework. Use decorator to build your model and map data.
  - name: keywords
    content: class model decorators ts.ed express typescript node.js javascript jsonschema json mapper serialization deserialization
---

# JsonMapper

The `@tsed/json-mapper` package is responsible to map a plain object to a model and a model to a plain object.

It provides two functions @@serialize@@ and @@deserialize@@ to transform object depending on which operation you want to perform.
It uses all decorators from `@tsed/schema` package and TypeScript metadata to work.

Ts.ED use this package to transform any input parameters sent by your consumer to a class and transform returned value by your endpoint
to a plain javascript object to your consumer.

## Configuration

```typescript
@Configuration({
  jsonMapper: {
    additionalProperties: false,
    disableUnsecureConstructor: false,
  }
})
```

### jsonMapper.additionalProperties

Enable additional properties on model. By default, `false`. Enable this option is dangerous and may be a potential security issue.

### jsonMapper.disableUnsecureConstructor

Pass the plain object to the model constructor. By default, `true`.

It may be a potential security issue if you have as constructor with this followings code:

```typescript
class MyModel {
  constructor(obj: any = {}) {
    Object.assign(this, obj); // potential prototype pollution
  }
}
```

## Usage

JsonMapper works with a class and decorators. Use decorators on properties to describe a model and use this model as an input parameter or return value by your endpoint. Here is a model example:

<Tabs class="-code">
  <Tab label="Person.ts">

<<< @/docs/snippets/converters/model-usage.ts

  </Tab>
  <Tab label="Jest">
  
<<< @/docs/snippets/converters/model-usage.jest.ts
  
  </Tab>
  <Tab label="Mocha">

<<< @/docs/snippets/converters/model-usage.mocha.ts

  </Tab>
</Tabs>

::: tip Note
Take a look on Jest/Mocha tabs to see @@serialize@@ and @@deserialize@@ functions usage.
:::

Now we can use the `Person` model on a controller:

<<< @/docs/snippets/converters/controller-usage.ts

::: tip Note
In the previous example, we can see @@Returns@@ decorator usage.
In all case, Ts.ED infer the returned value and apply the correct transformation on your response.

@@Returns@@ decorator is used to generate the correct swagger documentation only.
:::

::: warning
When a model is provided, JsonMapper will follow exactly the JsonSchema generated by `@tsed/schema` package.

It means, if you missed decorating one or more properties on your model, these properties won't be appear after the transformation.

<Tabs class="-code">
  <Tab label="User.ts">

<<< @/docs/snippets/converters/model-missing-properties.ts
</Tab>
<Tab label="Jest">

<<< @/docs/snippets/converters/model-missing-properties.jest.ts

  </Tab>
  <Tab label="Mocha">

<<< @/docs/snippets/converters/model-missing-properties.mocha.ts

  </Tab>
</Tabs>

> Note: Result is displayed in Jest/Mocha tabs.

:::

## Ignore properties (deprecated)

::: warning deprecated
This decorator is deprecated. Use @@Groups@@ decorator instead of.
:::

### Usage

@@Ignore@@ decorator can be used to ignore explicitly a property when a transformation have been performed.

For example, you have a base model to create a User named `UserCreation` where the `password` is required, but
you don't want to expose this field in other cases. One of the solution is to use class inheritance to solve this problem.

<Tabs class="-code">
  <Tab label="User.ts">

<<< @/docs/snippets/converters/model-ignore-props.ts

  </Tab>
  <Tab label="Jest">
  
<<< @/docs/snippets/converters/model-ignore-props.jest.ts
  
  </Tab>
  <Tab label="Mocha">

<<< @/docs/snippets/converters/model-ignore-props.mocha.ts

  </Tab>
</Tabs>

### With a callback

@@Ignore@@ decorator since v6.13.0 accept a callback which will be called when a property have been serialized or deserialized.
The callback will give you more control over the way to ignore a property.

```typescript
class User {
  @Name("id")
  _id: string;

  @Property()
  firstName: string;

  @Property()
  lastName: string;

  @Ignore((value, ctx) => ctx.endpoint) // should not serialized when the object is returned by an endpoint.
  password: string;

  @Ignore((value, ctx) => ctx.mongoose) // should be serialized when the object is returned by an endpoint.
  scopes: string[];

  @Ignore()
  alwaysIgnored: string;
}
```

Here is the available options on ctx:

| Prop     | Type    | Description              |
| -------- | ------- | ------------------------ |
| endpoint | boolean | It's an endpoint context |
| mongoose | boolean | It's a mongoose context  |

## Additional properties

@@AdditionalProperties@@ decorator can be used to accept any additional properties on a specific model.

<Tabs class="-code">
  <Tab label="Person.ts">

<<< @/docs/snippets/converters/model-additional-props.ts

  </Tab>
  <Tab label="Jest">
  
<<< @/docs/snippets/converters/model-additional-props.jest.ts
  
  </Tab>
  <Tab label="Mocha">

<<< @/docs/snippets/converters/model-additional-props.mocha.ts

  </Tab>
</Tabs>

## Alias

@@Name@@ decorator lets you to rename the exposed property in your json schema.

For example mongo db uses the `_id` property.
In order not to give any indication to our consumer about the nature of the database, it's better to rename the property to `id`.

<<< @/docs/snippets/model/name.ts

## OnSerialize

@@OnSerialize@@ decorator can be used to intercept and change the property value when a serialization is performed on class.

```typescript
import {OnSerialize} from "@tsed/schema";

export class Person {
  @OnSerialize((v) => v + "Test")
  property: string;
}
```

## OnDeserialize

@@OnDeserialize@@ decorator can be used to intercept and change the property value when a deserialization is performed on class.

```typescript
import {OnDeserialize} from "@tsed/schema";

export class Person {
  @OnDeserialize((v) => v + "Test")
  property: string;
}
```

## Type mapper

`@tsed/json-mapper` use classes to transform an input value to the expected value:

| Type       | Mapper               |
| ---------- | -------------------- |
| Primitives | @@PrimitiveMapper@@, |
| Symbol     | @@SymbolMapper@@,    |
| Objects    | @@DateMapper@@,      |

It's possible to add your own type mapper by using the @@JsonMapper@@ decorator on a class. Just copy a mapper implementation
and import the mapper in your application.

### Primitives

@@PrimitiveMapper@@ is responsible to map the primitive value like `Boolean`, `Number` or `String`.

<Tabs class="-code">
  <Tab label="PrimitiveMapper">
  
<<< @/../packages/specs/json-mapper/src/components/PrimitiveMapper.ts
  
  </Tab>

  <Tab label="Spec">
 
<<< @/../packages/specs/json-mapper/src/components/PrimitiveMapper.spec.ts
  
  </Tab>
</Tabs>

#### Cheat sheet

| Input       | Type    | Output                                                                             |
| ----------- | ------- | ---------------------------------------------------------------------------------- |
| `1`         | String  | `"1"`                                                                              |
| `"1"`       | String  | `"1"`                                                                              |
| `null`      | Number  | `null`                                                                             |
| `"null"`    | Number  | `null`                                                                             |
| `"1"`       | Number  | `1`                                                                                |
| `1`         | Number  | `1`                                                                                |
| `"to1"`     | Number  | Throw Bad Request. This is the only case where JsonMapper throw a cast type error. |
| `true`      | Boolean | `true`                                                                             |
| `"true"`    | Boolean | `true`                                                                             |
| `"1"`       | Boolean | `true`                                                                             |
| `1`         | Boolean | `true`                                                                             |
| `false`     | Boolean | `false`                                                                            |
| `"false"`   | Boolean | `false`                                                                            |
| `"0"`       | Boolean | `false`                                                                            |
| `0`         | Boolean | `false`                                                                            |
| `""`        | Boolean | `false`                                                                            |
| `"null"`    | Boolean | `null`                                                                             |
| `undefined` | Boolean | `undefined`                                                                        |

### Symbol

@@SymbolMapper@@ is responsible to map a `String` to `Symbol` or a `Symbol` to a `String`.

<Tabs class="-code">
  <Tab label="SymbolMapper">
  
<<< @/../packages/specs/json-mapper/src/components/SymbolMapper.ts

  </Tab>
  <Tab label="Spec">
 
<<< @/../packages/specs/json-mapper/src/components/SymbolMapper.spec.ts
  
  </Tab>
</Tabs>

### Date

@@DateMapper@@ is responsible to map a `Number`, `String` to a `Date` or a `Date` to a `String`.

<Tabs class="-code">
  <Tab label="DateMapper">

<<< @/../packages/specs/json-mapper/src/components/DateMapper.ts

  </Tab>
  <Tab label="Spec">

<<< @/../packages/specs/json-mapper/src/components/DateMapper.spec.ts

  </Tab>
</Tabs>

::: warning

Ts.ED doesn't transform Date to date format or hours format because it depends on each project guidelines.

But you can easily implement a Date mapper for each format with the Date API or moment:

```typescript
import {isBoolean} from "@tsed/core";
import {DateFormat} from "@tsed/schema";
import {serialize, JsonMapper, JsonMapperContext, JsonMapperMethods} from "../../src/index";

@JsonMapper(Date)
export class DateMapper implements JsonMapperMethods {
  deserialize(data: string | number, ctx: JsonMapperContext): Date;
  deserialize(data: boolean | null | undefined, ctx: JsonMapperContext): boolean | null | undefined;
  deserialize(data: any, ctx: JsonMapperContext): any {
    // don't convert unexpected data. In normal case, Ajv reject unexpected data.
    // But by default, we have to skip data deserialization and let user to apply
    // the right mapping
    if (isBoolean(data) || data === null || data === undefined) {
      return data;
    }

    return new Date(data);
  }

  serialize(object: Date, ctx: JsonMapperContext): any {
    const date = new Date(object);

    switch (ctx.options.format) {
      case "date":
        const y = date.getUTCFullYear();
        const m = ("0" + (date.getUTCMonth() + 1)).slice(-2);
        const d = ("0" + date.getUTCDate()).slice(-2);

        return `${y}-${m}-${d}`;
      default:
        return new Date(object).toISOString();
    }
  }
}
```

:::

## Create your own type mapper

It's possible de to change add your own type mapper by using the @@JsonMapper@@ decorator on a class. Just copy a mapper implementation
and import the mapper in your application.

A mapper must declare the type it must work on and implement two methods: serialize and deserialize.

```typescript
import {JsonMapper, JsonMapperMethods, JsonMapperCtx} from "@tsed/json-mapper";

@JsonMapper(String)
export class TheTypeMapper implements JsonMapperMethods {
  deserialize(data: any, ctx: JsonMapperCtx): String {
    return JSON.stringify(data) + ":deserialize";
  }

  serialize(data: any, ctx: JsonMapperCtx): String {
    return JSON.stringify(data) + ":serialize";
  }
}
```

Then import your new mapper in your Server.ts as following:

```typescript
import {Configuration} from "@tsed/di";

import "./mappers/TheTypeMapper";

@Configuration({
  mount: {
    "/rest": []
  }
})
export class Server {}
```

### Moment

[Moment.js](https://momentjs.com) is a powerful library to transform any formatted date string to a Moment instance.

You can change the Date mapper behavior to transform string to a Moment instance.

<Tabs class="-code">
  <Tab label="MomentMapper">
  
<<< @/docs/snippets/converters/moment-mapper.ts

  </Tab>
  <Tab label="Configuration">

<<< @/docs/snippets/converters/server-moment-mapper-import.ts

  </Tab>
  <Tab label="Example">

```typescript
import {Moment} from "moment";
import {Property} from "@tsed/schema";

export class Person {
  @Property(Date) // or @Property(String) + @DateTime()
  birthdate: Moment;
}
```

  </Tab>
</Tabs>

```

```
