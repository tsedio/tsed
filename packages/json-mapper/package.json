{
  "name": "@tsed/json-mapper",
  "version": "6.64.2",
  "description": "Json mapper module for Ts.ED Framework",
  "source": "./src/index.ts",
  "main": "./lib/index.js",
  "typings": "./lib/index.d.ts",
  "private": false,
  "keywords": [
    "TypeScript",
    "decorators",
    "models",
    "json mapper",
    "mapper",
    "serialization",
    "deserialization",
    "class",
    "classes",
    "tsed"
  ],
  "scripts": {
    "build": "tsc --build tsconfig.compile.json"
  },
  "dependencies": {
    "@tsed/core": "6.64.2",
    "tslib": "2.2.0"
  },
  "devDependencies": {
    "@tsed/core": "6.55.2",
    "@tsed/exceptions": "6.64.2",
    "@tsed/schema": "6.64.2"
  },
  "peerDependencies": {
    "@tsed/core": "^6.55.2",
    "@tsed/exceptions": "^6.55.2",
    "@tsed/schema": "^6.55.2"
  }
}