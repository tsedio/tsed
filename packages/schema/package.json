{
  "name": "@tsed/schema",
  "version": "6.64.2",
  "description": "JsonSchema module for Ts.ED Framework",
  "private": false,
  "source": "./src/index.ts",
  "main": "./lib/index.js",
  "typings": "./lib/index.d.ts",
  "exports": {
    "require": "./lib/index.js",
    "default": "./lib/index.modern.js"
  },
  "keywords": [
    "TypeScript",
    "decorators",
    "models",
    "json schema",
    "JsonSchema",
    "class",
    "classes",
    "tsed"
  ],
  "scripts": {
    "build": "microbundle --target node --no-compress --format modern,cjs --tsconfig ./tsconfig.compile.json"
  },
  "dependencies": {
    "@tsed/core": "6.64.2",
    "@tsed/openspec": "6.64.2",
    "change-case": "4.1.2",
    "micromatch": "4.0.2",
    "tslib": "2.2.0"
  },
  "devDependencies": {
    "@apidevtools/swagger-parser": "10.0.2",
    "@types/fs-extra": "9.0.6",
    "@types/json-schema": "7.0.7",
    "@types/micromatch": "4.0.1",
    "fs-extra": "9.0.1"
  },
  "peerDependencies": {
    "@tsed/core": "6.60.1",
    "@tsed/openspec": "6.60.1"
  }
}