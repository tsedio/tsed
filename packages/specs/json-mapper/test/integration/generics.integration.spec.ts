import {boolean, date, GenericOf, Generics, getJsonSchema, number, Property, string} from "@tsed/schema";
import {deserialize} from "@tsed/json-mapper";
import {expect} from "chai";

describe("Generics", () => {
  describe("using Functional api", () => {
    it("should transform string from generic object", () => {
      @Generics("T")
      class UserProperty<T> {
        @Property("T")
        value: T;
      }

      class Adjustment {
        @GenericOf(string())
        adjustment: UserProperty<string>;
      }

      let result = deserialize({
        adjustment: {
          value: "2019-01-01T12:45:57.111Z"
        }
      }, {
        type: Adjustment
      });

      expect(result).to.deep.eq({
        "adjustment": {
          "value": "2019-01-01T12:45:57.111Z"
        }
      });
    });
    it("should transform number from generic object", () => {
      @Generics("T")
      class UserProperty<T> {
        @Property("T")
        value: T;
      }

      class Adjustment {
        @GenericOf(number())
        adjustment: UserProperty<number>;
      }

      let result = deserialize({
        adjustment: {
          value: 10
        }
      }, {
        type: Adjustment
      });

      expect(result).to.deep.eq({
        "adjustment": {
          "value": 10
        }
      });
    });
    it("should transform boolean from generic object", () => {
      @Generics("T")
      class UserProperty<T> {
        @Property("T")
        value: T;
      }

      class Adjustment {
        @GenericOf(boolean())
        adjustment: UserProperty<Date>;
      }

      let result = deserialize({
        adjustment: {
          value: false
        }
      }, {
        type: Adjustment
      });

      expect(result).to.deep.eq({
        "adjustment": {
          "value": false
        }
      });
    });
    it("should transform date from generic object", () => {
      @Generics("T")
      class UserProperty<T> {
        @Property("T")
        value: T;
      }

      class Adjustment {
        @GenericOf(date().format("date-time"))
        adjustment: UserProperty<Date>;
      }

      let result = deserialize({
        adjustment: {
          value: "2019-01-01T12:45:57.111Z"
        }
      }, {
        type: Adjustment
      });

      expect(result).to.deep.eq({
        "adjustment": {
          "value": new Date("2019-01-01T12:45:57.111Z")
        }
      });
    });
  });
  describe("using type", () => {
    it("should transform string from generic object", () => {
      @Generics("T")
      class UserProperty<T> {
        @Property("T")
        value: T;
      }

      class Adjustment {
        @GenericOf(String)
        adjustment: UserProperty<string>;
      }

      let result = deserialize({
        adjustment: {
          value: "2019-01-01T12:45:57.111Z"
        }
      }, {
        type: Adjustment
      });

      expect(result).to.deep.eq({
        "adjustment": {
          "value": "2019-01-01T12:45:57.111Z"
        }
      });
    });
    it("should transform number from generic object", () => {
      @Generics("T")
      class UserProperty<T> {
        @Property("T")
        value: T;
      }

      class Adjustment {
        @GenericOf(Number)
        adjustment: UserProperty<number>;
      }

      let result = deserialize({
        adjustment: {
          value: 10
        }
      }, {
        type: Adjustment
      });

      expect(result).to.deep.eq({
        "adjustment": {
          "value": 10
        }
      });
    });
    it("should transform boolean from generic object", () => {
      @Generics("T")
      class UserProperty<T> {
        @Property("T")
        value: T;
      }

      class Adjustment {
        @GenericOf(Boolean)
        adjustment: UserProperty<Date>;
      }

      let result = deserialize({
        adjustment: {
          value: false
        }
      }, {
        type: Adjustment
      });

      expect(result).to.deep.eq({
        "adjustment": {
          "value": false
        }
      });
    });
    it("should transform date from generic object", () => {
      @Generics("T")
      class UserProperty<T> {
        @Property("T")
        value: T;
      }

      class Adjustment {
        @GenericOf(Date)
        adjustment: UserProperty<Date>;
      }

      let result = deserialize({
        adjustment: {
          value: "2019-01-01T12:45:57.111Z"
        }
      }, {
        type: Adjustment
      });

      expect(result).to.deep.eq({
        "adjustment": {
          "value": new Date("2019-01-01T12:45:57.111Z")
        }
      });
    });
    it("should generate JsonSchema with 'enum' from generic object", () => {
      enum AdjustmentType {
        PRICE = "price",
        DELAY = "delay",
      }

      @Generics("T")
      class UserProperty<T> {
        @Property("T")
        value: T;
      }

      class Adjustment {
        @GenericOf(AdjustmentType)
        adjustment: UserProperty<AdjustmentType>;
      }

      let result = deserialize({
        adjustment: {
          value: "delay"
        }
      }, {
        type: Adjustment
      });

      expect(result).to.deep.eq({
        "adjustment": {
          "value": "delay"
        }
      });
    });
  });
});