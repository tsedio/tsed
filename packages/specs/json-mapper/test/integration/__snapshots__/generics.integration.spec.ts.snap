// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generics > using Functional api > should transform date from generic object 1`] = `
Adjustment {
  "adjustment": UserProperty {
    "value": "2019-01-01T12:45:57.111Z",
  },
}
`;

exports[`Generics using Functional api should transform date from generic object 1`] = `
Adjustment {
  "adjustment": UserProperty {
    "value": "2019-01-01T12:45:57.111Z",
  },
}
`;
